swim/memberlist.go

```diff
@@ -389,69 +382,75 @@ func (m *memberlist) SetLocalLabels(labels map[string]string) error {
 // and subsequently be gossiped around. It is a valid operation to remove non-
 // existing keys. It returns true if all (and only all) labels have been removed.
 func (m *memberlist) RemoveLocalLabels(keys ...string) bool {
-	var before Member
-	m.members.Lock()
-	before = *m.local
-
-	if len(m.local.Labels) == 0 || len(keys) == 0 {
-		m.members.Unlock()
-		// nothing to delete
-		return false
-	}
+	// keep track if all labels are removed, it will be set to false if a label
+	// couldn't be removed.
+	removed := true
+
+	m.updateLocalMember(func(member *Member) bool {
+		// ensure that there is a new copy of the labels to work
+		// with.
+		labelsCopy := member.Labels.copy()
+
+		any := false // keep track if we at least removed one label
+		for _, key := range keys {
+			_, has := labelsCopy[key]
+			delete(labelsCopy, key)
+			removed = removed && has
+			any = any || has
+		}
 
-	// ensure that there is a new copy of the labels on the member to work with,
-	// CopyLabels guarantees a non-nil map so it is safe to work with
-	m.local.Labels = CopyLabels(m.local.Labels)
-
-	any := false    // keep track if we at least removed one label
-	removed := true // keep track if all labels are removed
-	for _, key := range keys {
-		_, has := m.local.Labels[key]
-		delete(m.local.Labels, key)
-		removed = removed && has
-		any = any || has
-	}
+		if any {
+			// only if there are changes we put the copied labels on the member.
+			member.Labels = labelsCopy
+		}
 
-	m.members.Unlock()
+		return any
+	})
 
-	if any {
-		// only reincarnate if there is a label removed
-		m.postLocalUpdate(before)
-	}
 	return removed
 }
 
-// postLocalUpdate should be called after the local Member has been updated to
-// make sure that its new state has a higher incarnation number and the change
-// will be recorded as a change to gossip around.
-func (m *memberlist) postLocalUpdate(before Member) {
-	// bump our incarnation for this change to be accepted by all peers
+// updateLocalMember takes an update function to upate the member passed in. The
+// update function can make mutations to the member and should indicate if it
+// made changes, only if changes are made the incarnation number will be bumped
+// and the new state will be gossiped to the peers
+func (m *memberlist) updateLocalMember(update func(*Member) bool) {
+	var changes []Change
+
 	m.members.Lock()
-	change := m.bumpIncarnation()
+
+	before := *m.local
+	didUpdate := update(m.local)
+	if didUpdate {
```

Comment:
We can change it to the early return, might be easier to understand.

However we cannot use defer here!

There are calls executed below that acquire their own lock which would result in a deadlock.